tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc

```diff
@@ -2035,6 +2099,16 @@ static IrArray::Index GetUnnormalizedIndex(
     const Shape& unnormalized_shape, llvm::IRBuilder<>* b_,
     const KernelMappingScheme& kernel_mapping_scheme) {
   DCHECK_EQ(normalized_shape_index.size(), 3);
+  // If the normalization only add a new dimensions of size 1,
+  // generate simpler indexing. LLVM doesn't always simplify the more
+  // complicated indexing and this prevent him from vectorizing some
+  // cases.
+  if (unnormalized_shape.rank() == 2) {
+    DCHECK_EQ(normalized_shape_index.dims()[0], 0);
+    auto multidim = normalized_shape_index.multidim();
+    return IrArray::Index({multidim[1], multidim[2]}, unnormalized_shape,
```

Comment:
Sorry, I am a bit confused on what is going on here, this code is new.
Why rank 2? Isn't `[2]` out of bounds in this case?

Reply:
Their is a "normalization" step of the input shapes to a 3d shape. So a matrix of shape MxN is converted to 1xMxN. A 3d tensor of shape MxNxK that we reduce on the first 2 dimensions is converted to a shape of 1xM*NxK, ...
This help simplify the code generator as the shapes he do not need to optimize those part.

But when we want to trigger the read, we need to "unnormalize" the shapes/indexing. This unnormalization isn't always well cleaned up by LLVM. This unnormalization is also done for each elements read, not once per threads.

So here, I simplify remove this unnormalization step when it isn't needed. This clean up the not optimized LLVM code and allow LLVM to vectorize more this case.

The unnormalization is taking the normalized 3d indexing, converting it to a 1d indexing, then converting it back to the original indexing. It is a afair amount of LLVM code.

as multidim is the normalized shapes that is always 3d, multidim[2] is valid.